### ISSUE TITLE
Lua `fiber:join()` is too unsafe

### ISSUE THREAD

**Issue by Gerold103** [tarantool/tarantool#7531]
### Issue 1
One example: https://github.com/tarantool/tarantool/issues/7489.

### Issue 2
Another notably worse example below.

* Create a file `test.c` with the following code:
```C
#include "module.h"
#include "lua.h"
#include "lauxlib.h"

#include <assert.h>

struct my_fiber_ctx {
	lua_State *stack;
	uint64_t id;
};

static int
my_fiber_f(va_list va)
{
	struct my_fiber_ctx *ctx = va_arg(va, struct my_fiber_ctx *);
	int rc = luaL_dostring(ctx->stack, "return require('fiber').self():id()");
	assert(rc == 0);
	(void)rc;
	ctx->id = luaL_touint64(ctx->stack, -1);
	ctx = NULL;

	while (!fiber_is_cancelled()) {
		say_info("alive");
		fiber_yield();
	}
	say_info("dead");
	return 0;
}

static int
my_fiber_start(lua_State *L)
{
	struct fiber_attr *attr = fiber_attr_new();
	fiber_attr_setstacksize(attr, 1024 * 1024);
	struct fiber *f = fiber_new_ex("my_fiber", attr, my_fiber_f);
	fiber_attr_delete(attr);
	assert(f != NULL);
	(void)f;
	fiber_set_joinable(f, true);

	struct my_fiber_ctx ctx;
	ctx.stack = lua_newthread(L);
	ctx.id = 0;
	fiber_start(f, &ctx);
	assert(ctx.id != 0);
	luaL_pushuint64(L, ctx.id);
	return 1;
}

LUA_API int
luaopen_test(lua_State *L)
{
	static const struct luaL_Reg lib[] = {
		{"my_fiber_start", my_fiber_start},
		{NULL, NULL},
	};
	luaL_register(L, "test", lib);
	return 1;
}
```
* Compile it into a shared library. On Mac I do it in `tarantool/sandbox` folder as `clang -g -dynamiclib -undefined dynamic_lookup -I../src/ -I../third_party/luajit/src/ -I../src/lib/msgpuck/ test.c -o test.dylib`.
* Run tarantool with `test.dylib` in the work dir.
* Execute the following Lua code:
```Lua
fiber = require('fiber')
test = require('test')
id = test.my_fiber_start()
f = fiber.find(id)

function join_f()
    print(f:join())
end
f1 = fiber.new(join_f) f2 = fiber.new(join_f)
f:cancel()
```
It crashes because of use-after-free with the following trace (but can be anything else easily):
```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
    frame #0: 0x00000001002baca1 tarantool`rlist_add_tail(head=0x00000001048006b0, item=0x00000001048009b0) at rlist.h:87:19
   84  	{
   85  		item->next = head;
   86  		item->prev = head->prev;
-> 87  		item->prev->next = item;
   88  		item->next->prev = item;
   89  	}
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
  * frame #0: 0x00000001002baca1 tarantool`rlist_add_tail(head=0x00000001048006b0, item=0x00000001048009b0) at rlist.h:87:19
    frame #1: 0x00000001002ba88f tarantool`fiber_join_timeout_internal(fiber=0x00000001048005c8, timeout=3153600000, retval=0x0000000104bffd2c) at fiber.c:645:4
    frame #2: 0x00000001002893c2 tarantool`lbox_fiber_join(L=0x00000001036e6240) at fiber.c:750:18
    frame #3: 0x000000010031b1fb tarantool`lj_BC_FUNCC + 68
    frame #4: 0x0000000100329241 tarantool`lua_pcall(L=0x00000001036e6240, nargs=0, nresults=-1, errfunc=0) at lj_api.c:1163:12
    frame #5: 0x0000000100291573 tarantool`luaT_call(L=0x00000001036e6240, nargs=0, nreturns=-1) at utils.c:471:6
    frame #6: 0x000000010028a9d4 tarantool`lua_fiber_run_f(ap=0x00000001048009f8) at fiber.c:378:11
```
In the test I created a fiber with a custom stack so as the `struct fiber` object would be deleted. I also deliberately didn't use `fiber.info()` to lookup my C-fiber by name. I wanted to show that I can fetch fiber ID from C to find it without fullscan.

### Issue 3
Another issue, with just Lua. And even after #7513 is applied:
```Lua
fiber = require('fiber')
f = nil
f1 = nil
f2 = nil
f3 = nil
block1 = true
function test_func1()
    while block1 do fiber.sleep(0.001) end
    print('test_func1')
    return 1
end
function test_func2()
    print('test_func2')
    return 2
end
function join_f()
    local res1, res2 = f:join()
    print(res1, res2)
    if res2 == 1 then
        f3 = fiber.new(test_func2)
        f3:set_joinable(true)
    end
end

f = fiber.new(test_func1)
f:set_joinable(true)
f1 = fiber.new(join_f)
f2 = fiber.new(join_f)
block1 = false
```
This crashes in the following way:
```
Assertion failed: (idx != 0 && -idx <= L->top - L->base), function index2adr, file /Users/gerold/Work/Repositories/tarantool/third_party/luajit/src/lj_api.c, line 40.
Process 29966 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = hit program assert
    frame #4: 0x000000010031edf2 tarantool`index2adr(L=0x0000000103298d98, idx=-1) at lj_api.c:40:5
   37  	    TValue *o = L->base + (idx - 1);
   38  	    return o < L->top ? o : niltv(L);
   39  	  } else if (idx > LUA_REGISTRYINDEX) {
-> 40  	    api_check(L, idx != 0 && -idx <= L->top - L->base);
   41  	    return L->top + idx;
   42  	  } else if (idx == LUA_GLOBALSINDEX) {
   43  	    TValue *o = &G(L)->tmptv;
* thread #1, queue = 'com.apple.main-thread', stop reason = hit program assert
    frame #0: 0x00007fff6bd5433a libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff6be10e60 libsystem_pthread.dylib`pthread_kill + 430
    frame #2: 0x00007fff6bcdb808 libsystem_c.dylib`abort + 120
    frame #3: 0x00007fff6bcdaac6 libsystem_c.dylib`__assert_rtn + 314
  * frame #4: 0x000000010031edf2 tarantool`index2adr(L=0x0000000103298d98, idx=-1) at lj_api.c:40:5
    frame #5: 0x0000000100320afb tarantool`lua_tointeger(L=0x0000000103298d98, idx=-1) at lj_api.c:399:16
    frame #6: 0x00000001002893ff tarantool`lbox_fiber_join(L=0x00000001032a4a70) at fiber.c:754:14
    frame #7: 0x000000010031b1fb tarantool`lj_BC_FUNCC + 68
    frame #8: 0x0000000100329241 tarantool`lua_pcall(L=0x00000001032a4a70, nargs=0, nresults=-1, errfunc=0) at lj_api.c:1163:12
    frame #9: 0x0000000100291573 tarantool`luaT_call(L=0x00000001032a4a70, nargs=0, nreturns=-1) at utils.c:471:6
```
I used the fact that the fibers are reused. Can even happen with a fiber who is being joined by 2 and more other fibers at the same time.

### Issue 4
Ok, assume in the issue 3 we managed to fix it by never freeing the fibers. Then the second join would try to join a totally unrelated new fiber. Simply because `struct fiber` was reused. Look at the slightly modified example:
```Lua
fiber = require('fiber')
f = nil
f1 = nil
f2 = nil
f3 = nil
block1 = true
function test_func1()
    while block1 do fiber.sleep(0.001) end
    print('test_func1')
    return 1
end
function test_func2()
    print('test_func2')
    while true do
        fiber.sleep(1000000)
    end
    return 2
end
function join_f()
    local res1, res2 = f:join()
    print(res1, res2)
    if res2 == 1 then
        f3 = fiber.new(test_func2)
        f3:set_joinable(true)
    end
end

f = fiber.new(test_func1)
f:set_joinable(true)
f1 = fiber.new(join_f)
f2 = fiber.new(join_f)
block1 = false
```
`f2` never ends. Both `f1` and `f2` were trying to join `f1`, but `f2` is still hanging. This is because inside of C API `fiber_join()` the fiber `f`'s `struct fiber` got reused. `f2` never noticed that `f1` is dead. It just keeps waiting until it will join the brand new unrelated fiber.

**Comment by Gumix** [comment]
I managed to reproduce Issue 3 and 4 only after the following change:
```diff
f = fiber.new(test_func1)
f:set_joinable(true)
- f1 = fiber.new(join_f)
- f2 = fiber.new(join_f)
+ f1 = fiber.create(join_f)
+ f2 = fiber.create(join_f)
block1 = false
```
